### ISSUE TITLE
hs.httpserver and ssl

### ISSUE THREAD

**Issue by asmagill** [Hammerspoon/hammerspoon#539]
Currently when I try to visit a hammerspoon httpserver that has SSL enabled, it times out or gives me a "cannot create a secure connection" message in Safari... is this an artifact of running a development build  of Hammerspoon or something I should dig into deeper to see what is not working as expected?

**Comment by cmsj** [comment]
How are you finding/opening the URL in Safari? I ask because if you're using the Bonjour section of Bookmarks, that defaults to using http:// so you'd need to manually add https:// to the URL.

**Comment by asmagill** [comment]
I've also typed it in by hand and made sure the https was included.  Here's a screen shot and the code I'm using to test with:

![screen shot 2015-09-08 at 12 58 51 am](https://cloud.githubusercontent.com/assets/8139480/9727473/ed3b97c0-55c4-11e5-8d9d-5091928254db.png)

~~~lua
h = hs.httpserver.new(true):setName("test"):setCallback(function(t,p) return [[
<html>
  <head><title>Yay!</title></head>
  <body bgcolor="#cccc00">
      You made it with ]]..t..[[ ]]..p..[[!<hr>
      <div align="right"><i>]]..os.date()..[[</i></div>
  </body>
</html>]], 200, {} end):start()
~~~

I've tried with localhost, my machine name on the network, and by ip address (127.0.0.1 and the one assigned to my wifi).  At first I thought it might be because I also had a password set in my first tests, but I took that out as you can see.  And the password works fine when it's not using ssl.

**Comment by cmsj** [comment]
Odd, that works for me, using https://localhost:<port>
I'll push up a patch that enables CocoaHTTPServer logging to Console.app

**Comment by cmsj** [comment]
(pushed as c2bc7b7)

**Comment by cmsj** [comment]
Also, does it with with http?

**Comment by asmagill** [comment]
It works fine as regular http (when I leave the true out), but with true, none of them work.  I'll give it a shot here in a minute.  Thanks!

**Comment by asmagill** [comment]
~~~
// started hs.httpserver web server

9/8/15 1:34:04.158 AM launchservicesd[84]: SecTaskLoadEntitlements failed error=22
9/8/15 1:34:04.549 AM Hammerspoon[65254]: HTTPServer: Started HTTP server on port 54601
9/8/15 1:34:05.443 AM Hammerspoon[65254]: Bonjour Service Published: domain(local.) type(_http._tcp.) name(test)
9/8/15 1:34:08.090 AM launchservicesd[84]: SecTaskLoadEntitlements failed error=22

// switched to safari and tried to load URL

9/8/15 1:34:32.726 AM pkd[45834]: SecTaskLoadEntitlements failed error=22
9/8/15 1:34:32.751 AM launchservicesd[84]: SecTaskLoadEntitlements failed error=22
9/8/15 1:34:34.239 AM launchservicesd[84]: SecTaskLoadEntitlements failed error=22
9/8/15 1:34:38.982 AM WindowServer[196]: _CGXRemoveWindowFromWindowMovementGroup: window 0x259e is not attached to window 0x25dd
9/8/15 1:34:41.242 AM com.apple.WebKit.Networking[52109]: CFNetwork SSLHandshake failed (-9801)
9/8/15 1:34:41.276 AM com.apple.WebKit.Networking[52109]: CFNetwork SSLHandshake failed (-9801)
9/8/15 1:34:41.285 AM com.apple.WebKit.Networking[52109]: CFNetwork SSLHandshake failed (-9801)
9/8/15 1:34:41.286 AM com.apple.WebKit.Networking[52109]: NSURLSession/NSURLConnection HTTP load failed (kCFStreamErrorDomainSSL, -9801)

// restart Hammerspoon and try again...

9/8/15 1:41:16.237 AM Hammerspoon[65953]: HTTPServer: Started HTTP server on port 54623
9/8/15 1:41:17.014 AM Hammerspoon[65953]: Bonjour Service Published: domain(local.) type(_http._tcp.) name(test2)
9/8/15 1:41:18.929 AM launchservicesd[84]: SecTaskLoadEntitlements failed error=22
9/8/15 1:41:22.947 AM launchservicesd[84]: SecTaskLoadEntitlements failed error=22


9/8/15 1:41:45.397 AM pkd[45834]: SecTaskLoadEntitlements failed error=22
9/8/15 1:41:45.493 AM launchservicesd[84]: SecTaskLoadEntitlements failed error=22
9/8/15 1:41:50.159 AM launchservicesd[84]: SecTaskLoadEntitlements failed error=22
9/8/15 1:41:55.559 AM WindowServer[196]: _CGXRemoveWindowFromWindowMovementGroup: window 0x259a is not attached to window 0x3641
9/8/15 1:41:57.126 AM com.apple.WebKit.Networking[52109]: CFNetwork SSLHandshake failed (-9801)
9/8/15 1:41:57.160 AM com.apple.WebKit.Networking[52109]: CFNetwork SSLHandshake failed (-9801)
9/8/15 1:41:57.170 AM com.apple.WebKit.Networking[52109]: CFNetwork SSLHandshake failed (-9801)
9/8/15 1:41:57.171 AM com.apple.WebKit.Networking[52109]: NSURLSession/NSURLConnection HTTP load failed (kCFStreamErrorDomainSSL, -9801)
~~~

I recently upgraded to 10.11 beta... do I need to change a setting somewhere to get prompted about self-signed certificates?  I use a few for various other sites, but they were approved long before the upgrade so I no longer get prompted for them and I don't think I've needed to add any since the upgrade.

I also use LittleSnitch, but Hammerspoon is green-lit for anything in or out, and it didn't show any warnings... and turning off it's Network Filter doesn't make a difference... I can try completely uninstalling it tomorrow if you think it's worth the effort.

I'll also try a release version later as well and see if that makes a difference... I'm pretty sure the only differences I have in place right now are the LuaSkin WIP I'm testing out and I haven't applied the changes you've added today yet, but none of those *should* make a difference... and the webview module I'm working on is currently an external module (and not loaded presently at that) -- the branch I've pushed for review isn't my active branch right now.

At any rate, unless any of this sparks a thought of something you want me to try, I won't worry too much about it until I've had a chance to try it out with the release version of Hammerspoon.

**Comment by cmsj** [comment]
Interesting. I'll give it a go on my 10.11 machine later. I'm not aware of anything that should change an app's ability to listen on sockets with TLS.

**Comment by cmsj** [comment]
ok, I can reproduce this with 0.9.39 on 10.11. It looks to me like CocoaAsyncSocket is not doing TLS1.2, so App Transport Security is refusing to talk to it.

I hope we can just get away with updating CocoaAsyncSocket (we're two point releases behind) and/or rejigging the TLS/SSL settings we're using, but I do have a longer term fear about relying on three cocoapods to provide httpserver :/

**Comment by cmsj** [comment]
Found it. I had to override a method in CocoaHTTPServer to not use some old deprecated SSL defines, and for some reason I hard coded it to use SSL 3.0. From 10.11 onwards, that's deemed to be very bad, so it is now forced to use the most secure protocol, which is TLS1.2 (which is available from 10.8 onwards).
